### DISCUSSION TITLE
Content Delivery API - There should be an easy option to enable Swagger in non-development mode

### DISCUSSION THREAD

**Discussion by lauraneto** [umbraco/Umbraco-CMS#14393] [Features and ideas]
As we are moving towards headless, having API management (Azure API management, for example) in front of Umbraco will become a more common scenario.

In order to import the endpoint configuration into the API management instance, at my company we usually run a task after deployment to read the swagger specification and import the endpoints configuration automatically (as part of the CI/CD).

At the moment, having the swagger disabled in non Development mode is enforced and while you can do workarounds for it, from what I understood, I do believe there is a valid use case to make this a setting (which can default to disabled).

In my case the web apps themselves will be protected/IP restricted (so no security risk there), and only the endpoints configured in the swagger are exposed. (the actual swagger specification isn't)

I look forward to hear your thoughts!

**Comment by vsilvar** [comment]
In my opinion, as a general rule of thumb, behavior shouldn't be based solely on the environment code is being run at, but instead by a specific setting that controls it.
It's very frustrating to have no way of overriding some of this behavior.
(Or having to copy/re-implement Umbraco code to slightly tweak something)

In this specific case, for example, I think an `Umbraco:CMS:DeliveryApi:Swagger` appsetting should exist, which could either be enabled, disabled or automatic (which would retain the current behavior).
This could be done with either an enum or a nullable boolean, where the null state is the automatic.

But this is also present in other places, from memory, I've also ran into the exact same issue when trying to disable Razor runtime compilation, on the local environment. You simply can't unless you also change the runtime mode to Production, which is not what I want.

So would be nice to have these both addressed, but also that this is something that is taken into account when new features are implemented.

Relevant code:
https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Cms.Api.Common/DependencyInjection/UmbracoBuilderApiExtensions.cs#L44
https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Web.Common/DependencyInjection/UmbracoBuilderExtensions.cs#L236

**Comment by Danielh21** [comment]
I agree it should be easy to turn on.
@lauraneto you mentioned some workarounds to enable swagger in production? Can you share a bit on how to do that, we might need it as well :)

**Comment by bergmania** [comment]
Sounds okay with me. We can include it in a minor version, if somebody are up for making a PR 💪
